### PR TITLE
FortiGate ban feature

### PR DESCRIPTION
**Pull request by guykeller**
## Status 
Ready

## Related Issues
fixes: https://github.com/demisto/etc/issues/17701

## Description
Added several commands to FortiGate - ban-ip, unban-ip, get-banned-ips.

## Does it break backward compatibility?
   - No

## Must have
- [ ] Documentation (with link to it)
- [ ] Code Review

### CODE REVIEW

File: Integrations/FortiGate/FortiGate.py [hunk 1]
```diff
@@ -94,6 +94,44 @@ def create_addr_string(list_of_addr_data_dicts):
...
+    try:
+        arg_int = int(arg_str)
+    except Exception:
+        return_error("Error: {0} must have an integer value.".format(arg_name_str))
+    return arg_int
+
+
+def prettify_date(date_string):
+    date_string = date_string[:-5]  # remove the .000z at the end
+    date_parts = date_string.split("T")
```
Comment (yaakovi) [inline Integrations/FortiGate/FortiGate.py]: why split and format when you can just replace?
```suggestion
    date_parts = date_string.replace('T', ' ')
```

File: Integrations/FortiGate/FortiGate.py [hunk 2]
```diff
@@ -94,6 +94,44 @@ def create_addr_string(list_of_addr_data_dicts):
...
+            expiration_in_ms = 1000 * int(ip_data.get("expires"))
+            current_ip_context["Expires"] = prettify_date(timestamp_to_datestring(expiration_in_ms))
+        if ip_data.get("created"):
+            creation_in_ms = 1000 * int(ip_data.get("created"))
+            current_ip_context["Created"] = prettify_date(timestamp_to_datestring(creation_in_ms))
+        ips_contexts_array.append(current_ip_context)
+    return ips_contexts_array
+
+
+def create_banned_ips_human_readable(entry_context, response):
```
Comment (yaakovi) [inline Integrations/FortiGate/FortiGate.py]: you are not using the `response` argument

File: Integrations/FortiGate/FortiGate.py [hunk 3]
```diff
@@ -429,6 +467,76 @@ def create_firewall_service_request(service_name, tcp_range, udp_range):
     return response
 
 
+def ban_ip(ip_addresses_array, time_to_expire):
```
Comment (yaakovi) [inline Integrations/FortiGate/FortiGate.py]: no need to specify `array`, using plural form is enough
```suggestion
def ban_ip(ip_addresses, time_to_expire):
```

File: Integrations/FortiGate/FortiGate.py [hunk 4]
```diff
@@ -429,6 +467,76 @@ def create_firewall_service_request(service_name, tcp_range, udp_range):
...
+    uri_suffix = 'monitor/user/banned/add_users/'
+
+    payload = {
+        'ip_addresses': ip_addresses_array
+    }
+    if time_to_expire:
+        time_to_expire_int = convert_arg_to_int(time_to_expire, "expiry")
```
Comment (yaakovi) [inline Integrations/FortiGate/FortiGate.py]: this should be handled in `ban_ip_command`.
the purpose of separating to 2 functions is: outer function will process the user input (handling arrays, numbers, dates, etc...) and inner function will handle the request logic.

here, you can use the value `0` or `None` for no expiration time

File: Integrations/FortiGate/FortiGate.py [hunk 5]
```diff
@@ -429,6 +467,76 @@ def create_firewall_service_request(service_name, tcp_range, udp_range):
...
+        payload["expiry"] = time_to_expire_int
+    response = http_request('POST', uri_suffix, {}, json.dumps(payload))
+    return response
+
+
+def ban_ip_command():
+    ip_addresses_string = demisto.args()["ip_addresses"]
+    ip_addresses_array = ip_addresses_string.split(",")
```
Comment (yaakovi) [inline Integrations/FortiGate/FortiGate.py]: use `argToList`

File: Integrations/FortiGate/FortiGate.yml [hunk 6]
```diff
@@ -437,6 +437,38 @@ script:
       description: Was address group deleted
       type: boolean
     description: Delete an address group from FortiGate firewall
+  - name: fortigate-ban-ip
+    arguments:
+    - name: ip_addresses
```
Comment (yaakovi) [inline Integrations/FortiGate/FortiGate.yml]: mark this argument with `isArray`

File: Integrations/FortiGate/FortiGate.yml [hunk 7]
```diff
@@ -437,6 +437,38 @@ script:
...
+      required: true
+      description: List of IP Addresses to ban, comma separated. IPv4 and IPv6 addresses
+        are allowed.
+    - name: expiry
+      description: Time until expiry in seconds. 0 for indefinite ban.
+    description: Immediately add one or more users to the banned list.
```
Comment (yaakovi) [inline Integrations/FortiGate/FortiGate.yml]: IPs, not users

File: Integrations/FortiGate/FortiGate.yml [hunk 8]
```diff
@@ -437,6 +437,38 @@ script:
...
+  - name: fortigate-unban-ip
+    arguments:
+    - name: ip_addresses
```
Comment (yaakovi) [inline Integrations/FortiGate/FortiGate.yml]: `isArray`

File: Integrations/FortiGate/FortiGate.yml [hunk 9]
```diff
@@ -437,6 +437,38 @@ script:
...
+      required: true
+      description: List of banned user IPs to clear, comma separated. IPv4 and IPv6
+        addresses are allowed.
+    description: Immediately clear a list of specific banned users by IP.
```
Comment (yaakovi) [inline Integrations/FortiGate/FortiGate.yml]: same here

File: Integrations/FortiGate/FortiGate.py [hunk 10]
```diff
@@ -429,6 +467,76 @@ def create_firewall_service_request(service_name, tcp_range, udp_range):
...
+    })
+
+
+def get_banned_ips():
+    uri_suffix = 'monitor/user/banned/select/'
+    response = http_request('GET', uri_suffix)
+    return response
+
+
+def get_banned_ips_command():
```
Comment (yaakovi) [inline Integrations/FortiGate/FortiGate.py]: what about adding those functions to the execution code (command switch-case)

File: Integrations/FortiGate/FortiGate.py [hunk 11]
```diff
@@ -429,6 +467,76 @@ def create_firewall_service_request(service_name, tcp_range, udp_range):
...
+    response = http_request('POST', uri_suffix, {}, json.dumps(payload))
+    return response
+
+
+def ban_ip_command():
+    ip_addresses_string = demisto.args()["ip_addresses"]
+    ip_addresses_array = ip_addresses_string.split(",")
+    time_to_expire = demisto.args().get("expiry")
+    response = ban_ip(ip_addresses_array, time_to_expire)
+    demisto.results({
```
Comment (yaakovi) [inline Integrations/FortiGate/FortiGate.py]: use `return_outputs`.
add context output (you already set the output path in `get_banned_ips`

File: Integrations/FortiGate/FortiGate.yml [hunk 12]
```diff
@@ -437,6 +437,38 @@ script:
...
+    - contextPath: Fortigate.BannedIP.Created
+      description: When this IP was added to the banned list.
+      type: string
+    - contextPath: Fortigate.BannedIP.Expires
+      description: When this IP will be expired from the banned list.
+      type: string
+    - contextPath: Fortigate.BannedIP.Source
+      description: Source of the ban.
+      type: string
+    description: Return a list of all banned users by IP.
```
Comment (yaakovi) [inline Integrations/FortiGate/FortiGate.yml]: same

**Comment by yaakovi** [comment]
need to update test playbook as well.

File: Integrations/FortiGate/FortiGate.py (hunk 3 above)
Comment (guykeller) [reply Integrations/FortiGate/FortiGate.py]: Could you please explain further?
How could someone reading the code be sure if the function is getting (and thus, sending in the request) an array or a string of multiple IPs, without actually having to read through the API or going back to the ban_ip_command function calling ban_ip?

File: Integrations/FortiGate/FortiGate.yml (hunk 7 above)
Comment (guykeller) [reply Integrations/FortiGate/FortiGate.yml]: It's taken directly from their documentation - IPs and users are the same according to them

File: Integrations/FortiGate/FortiGate.yml (hunk 9 above)
Comment (guykeller) [reply Integrations/FortiGate/FortiGate.yml]: same reply :-)

File: Integrations/FortiGate/FortiGate.yml (hunk 12 above)
Comment (guykeller) [reply Integrations/FortiGate/FortiGate.yml]: same :-)

File: Integrations/FortiGate/FortiGate.py (hunk 10 above)
Comment (guykeller) [reply Integrations/FortiGate/FortiGate.py]: Was written in the integration but forgotten on merge

File: Integrations/FortiGate/FortiGate.py (hunk 11 above)
Comment (guykeller) [reply Integrations/FortiGate/FortiGate.py]: AFAIK return_outputs doesn't handle well when you don't have any context outputs.
Moreover - we don't have context outputs by choice for these commands, it's a design choice verified with Arseny.

**Comment by guykeller** [comment]
There is no test playbook as there is no instance

File: Integrations/FortiGate/FortiGate.py (hunk 11 above)
Comment (yaakovi) [reply Integrations/FortiGate/FortiGate.py]: ok

File: Integrations/FortiGate/FortiGate.py [hunk 13]
```diff
@@ -94,6 +94,43 @@ def create_addr_string(list_of_addr_data_dicts):
     return addr_string
 
 
+def convert_arg_to_int(arg_str, arg_name_str):
+    try:
+        arg_int = int(arg_str)
+    except Exception:
```
Comment (Itay4) [inline Integrations/FortiGate/FortiGate.py]: `Exception` is too broad, catch `ValueError`

**Comment by yaakovi** [comment]
@guykeller 
this one: `TestPlaybooks/playbook-Fortigate_Test.yml`

File: Integrations/FortiGate/FortiGate.py [hunk 14]
```diff
@@ -94,6 +94,43 @@ def create_addr_string(list_of_addr_data_dicts):
...
+        return_error("Error: {0} must have an integer value.".format(arg_name_str))
+    return arg_int
+
+
+def prettify_date(date_string):
```
Comment (Itay4) [inline Integrations/FortiGate/FortiGate.py]: add docstring and elaborate about input and output

File: Integrations/FortiGate/FortiGate.py [hunk 15]
```diff
@@ -94,6 +94,43 @@ def create_addr_string(list_of_addr_data_dicts):
...
+
+def create_banned_ips_entry_context(ips_data_array):
+    ips_contexts_array = []
+    for ip_data in ips_data_array:
+        current_ip_context = {
+            "IP": ip_data.get("ip_address"),
+            "Source": ip_data.get("source")
+        }
+        if ip_data.get("expires"):
+            expiration_in_ms = 1000 * int(ip_data.get("expires"))
```
Comment (Itay4) [inline Integrations/FortiGate/FortiGate.py]: this might rasie `TypeError`, add default to `get()`

File: Integrations/FortiGate/FortiGate.py [hunk 16]
```diff
@@ -429,6 +466,79 @@ def create_firewall_service_request(service_name, tcp_range, udp_range):
...
 
+def ban_ip(ip_addresses_array, time_to_expire):
+    uri_suffix = 'monitor/user/banned/add_users/'
+
+    payload = {
+        'ip_addresses': ip_addresses_array,
+        'expiry': time_to_expire
+    }
+
+    response = http_request('POST', uri_suffix, {}, json.dumps(payload))
```
Comment (Itay4) [inline Integrations/FortiGate/FortiGate.py]: don't pass empty values like `{}`, add default values to `http_request`

File: Integrations/FortiGate/FortiGate.py [hunk 17]
```diff
@@ -429,6 +466,79 @@ def create_firewall_service_request(service_name, tcp_range, udp_range):
...
+
+
+def ban_ip_command():
+    ip_addresses_string = demisto.args()["ip_addresses"]
+    ip_addresses_array = argToList(ip_addresses_string)
+    time_to_expire = demisto.args().get("expiry")
+    if time_to_expire:
+        time_to_expire = convert_arg_to_int(time_to_expire, "expiry")
+    else:
+        time_to_expire = 0
```
Comment (Itay4) [inline Integrations/FortiGate/FortiGate.py]: add a comment to explain the logic for this default
also, set a default of `0` to the argument so the user will be aware of it

File: Integrations/FortiGate/FortiGate.yml [hunk 18]
```diff
@@ -437,6 +437,40 @@ script:
...
+      required: true
+      description: List of IP Addresses to ban, comma separated. IPv4 and IPv6 addresses
+        are allowed.
+      isArray: true
+    - name: expiry
+      description: Time until expiry in seconds. 0 for indefinite ban.
+    description: Immediately add one or more users to the banned list.
+  - name: fortigate-unban-ip
+    arguments:
+    - name: ip_addresses
```
Comment (Itay4) [inline Integrations/FortiGate/FortiGate.yml]: our convention is singular (the user can understand from the description that CSV is supported), please change to `ip_address`

File: Integrations/FortiGate/FortiGate.yml [hunk 19]
```diff
@@ -437,6 +437,40 @@ script:
...
+      required: true
+      description: List of banned user IPs to clear, comma separated. IPv4 and IPv6
```
Comment (Itay4) [inline Integrations/FortiGate/FortiGate.yml]: add example

File: Integrations/FortiGate/FortiGate.py (hunk 17 above)
Comment (Itay4) [reply Integrations/FortiGate/FortiGate.py]: @guykeller dont think you understood me here
add `0` as default to the argument in the integration settings

**Comment by guykeller** [comment]
@yaakovi Could you explain your request about the test playbook please?
@Itay4 :
Handled all the suggested changes. As for your other comments:
1. There are no possible errors.
Un-banning an IP that isn't banned will be successful as far as FortiGate is concerned.
The same goes for banning an IP that never communicated with your firewall, or banning an IP that is already banned.
2. According to the docs there is no difference, and I'm not familiar with any PEP8 input on this. 
Is it our convention? I'm not familiar with it, could you please point me to where it's written in case I missed anything else?
3. The decision to not change the context with the ban and unban command was made with Arseny.

**Comment by Itay4** [comment]
1. what are the outputs of `!fortigate-ban-ip ip_address="NotValidIPAddress"` and `!fortigate-ban-ip ip_address="0.0.0.0"` ?
2. this is our convention
3. dont see a reason not to, lets discuss so i can understand this better, perhaps im missing something

File: Integrations/FortiGate/FortiGate.py (hunk 17 above)
Comment (guykeller) [reply Integrations/FortiGate/FortiGate.py]: @Itay4 Gotcha now, will do

**Comment by guykeller** [comment]
Closing this and opening a new PR.

**Comment by yardensade** [comment]
Can we delete this branch?

**Pull request closed without merging**